pkg/certificates/certificates.go

```diff
@@ -0,0 +1,286 @@
+
+const (
+	// AuthorizationsBucket is the bucket used with a bolt-backed authorizations DB.
+	AuthorizationsBucket = "authorizations"
+	tokenLength          = 64 // 2^(64*8) =~ 1.34E+154
+)
+
+var (
+	mon = monkit.Package()
+	// ErrAuthorization is used when an error occurs involving an authorization.
+	ErrAuthorization = errs.Class("authorization error")
+	// ErrAuthorizationDB is used when an error occurs involving the authorization database.
+	ErrAuthorizationDB = errs.Class("authorization db error")
+	// ErrNotEnoughRandom is used when the number of bytes read from a random source are insufficient.
+	ErrNotEnoughRandom = ErrAuthorization.New("unable to read enough random bytes")
+	// ErrAuthorizationCount is used when attempting to create an invalid number of authorizations.
+	ErrAuthorizationCount = ErrAuthorizationDB.New("cannot add less than one authorizations")
+)
+
+// CertSignerConfig is a config struct for use with a certificate signing service
+type CertSignerConfig struct {
+	Overwrite          bool   `help:"if true, overwrites config AND authorization db is truncated" default:"false"`
+	AuthorizationDBURL string `help:"url to the certificate signing authorization database" default:"bolt://$CONFDIR/authorizations.db"`
+}
+
+// CertificateSigner implements pb.CertificatesServer
+type CertificateSigner struct {
+	Log *zap.Logger
+}
+
+// AuthorizationDB stores authorizations which may be claimed in exchange for a
+// certificate signature.
+type AuthorizationDB struct {
+	DB storage.KeyValueStore
+}
+
+// Authorizations is a slice of authorizations for convenient de/serialization
+// and grouping.
+type Authorizations []*Authorization
+
+// Authorization represents a single-use authorization token and it's status
+type Authorization struct {
+	Token Token
+	Claim *Claim
+}
+
+// Token is a random byte array to be used like a pre-shared key for claiming
+// certificate signatures.
+type Token [tokenLength]byte
+
+// Claim holds information about the circumstances under which an authorization
+// token was claimed.
+type Claim struct {
+	IP         string
+	Timestamp  int64
+	Identity   *provider.PeerIdentity
+	SignedCert *x509.Certificate
+}
+
+// NewServer creates a new certificate signing grpc server
+func NewServer(log *zap.Logger) pb.CertificatesServer {
+	srv := CertificateSigner{
+		Log: log,
+	}
+
+	return &srv
+}
+
+// NewAuthorization creates a new, unclaimed authorization with a random token value
+func NewAuthorization() (*Authorization, error) {
+	var token Token
+	i, err := rand.Read(token[:])
+	if err != nil {
+		return nil, ErrAuthorization.Wrap(err)
+	}
+	if i != tokenLength {
+		return nil, ErrNotEnoughRandom
+	}
+
+	return &Authorization{
+		Token: token,
+	}, nil
+}
+
+// NewAuthDB creates or opens the authorization database specified by the config
+func (c CertSignerConfig) NewAuthDB() (*AuthorizationDB, error) {
+	// TODO: refactor db selection logic?
+	driver, source, err := utils.SplitDBURL(c.AuthorizationDBURL)
```

Comment:
Can the database be using the masterdb as in the satellite/satellitedb interface? Like what we did for irreparabledb, statdb, bwagreements ..

Reply:
As we discussed on slack, copied here for posterity:

Yes it can but i'm not familiar with that part of the code yet, so in the interest of expediency I'd rather hold off on that for now. I'm also not sure how this will be deployed with respect to a "master database server".

AFAIK, this is only something we're doing for the alpha network. We're really not interested in being in the business of certificate authorities and cert-signing. There may be use cases for this service with our platform in other specific use cases but not in our general, long-term use case.